sphinxsearch: Add support for MySQL & xmlpipe2

Valodim opened the pull request:
###### Motivation for this change

This PR obsoletes #79032.

I added support for MySQL sources, and optionally xmlpipe2. I don't use xmlpipe2 myself, so I didn't test that. I'll test MySQL source in the next few days in a small production environment. :)

The version and main source were pulled in as arguments? Not sure what that was about, overriding those should be done with `overrideAttrs`. I also enabled parallel builds (works fine for me), and added myself as maintainer. The 2.X branch of sphinxsearch isn't actively supported anymore though, so I don't expect any updates. The 3.X line isn't open source (yet), and has enough incompatibilities to go into its own sphinxsearch3 package anyways.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] ~~Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`~~ no packages depend on this
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Valodim committed:
sphinxsearch: add support for mysql sources

Valodim committed:
sphinxsearch: add optional support for xmlpipe2

Valodim committed:
sphinxsearch: enable parallel builds

Valodim commented:
Closure size change (second is new):
> /nix/store/id1lrdwxwwlndsb5w3hr4v69zlhg5i4l-sphinxsearch-2.2.11	  57.2M
> /nix/store/6kdis2k84mfzdabmx5cfs6djy95zrk8g-sphinxsearch-2.2.11	  58.3M

Valodim commented:
Whoops. I built this against release-19.09 and it seems I didn't correctly rebase to master. will fix.

Valodim commented:
Ok, should work now. Closure sizes:
> sphinxsearch on master
> /nix/store/id1lrdwxwwlndsb5w3hr4v69zlhg5i4l-sphinxsearch-2.2.11	  57.2M
> sphinxsearch on pr, enableMysql = false
> /nix/store/8asbjxjxzbilhc25r1mj01zs2slhg71b-sphinxsearch-2.2.11	  57.2M
> sphinxsearch on pr, enableMysql = true (default)
> /nix/store/qjhabpy7dhvs8s3cmhzvzz2kkasvk6p2-sphinxsearch-2.2.11	  73.5M
> sphinxsearch on pr, enableMysql = true, enableXmlpipe2 = true
> /nix/store/z39nqslg1z21x3kc5smxcyqfqpp06ixk-sphinxsearch-2.2.11	  73.8M

Valodim commented:
Running in production, works just fine with a mysql source :+1:

Valodim commented:
Can I get a review on this one? :) @jtojnar @vvs-

jtojnar commented on pkgs/servers/search/sphinxsearch/default.nix:
```diff
@@ -1,31 +1,44 @@
-    "--without-mysql"
     "--enable-id64"
+  ] ++ (if enableMysql then [
+    "--with-mysql=${libmysqlclient}/"
```
Can this work with

```suggestion
    "--with-mysql"
```

vvs- replied:
As far as my previous PR goes configure didn't find anything in that case. And I don't see why it should be different here.

Valodim replied:
it'll actually find the lib without any parameter :) I changed it to just use `--without-mysql` if the feature's not enabled. The optionals reflect the default expected configuration in this way.

vvs- replied:
That's really weird, I'm absolutely positive that it didn't work previously and I've tested it many times before adding that variable. I can't explain this and I'd better retest your expression in my environment to avoid any surprises.

vvs- replied:
That's really weird, I'm absolutely positive that it didn't work previously and I've tested it many times before adding that variable. I can't explain this and I'd better retest your expression in my environment to avoid any surprises.

Valodim replied:
Yup, that's likely it. That also explains why you needed the explicit dependencies on openssl and zlib, which are propagated by libmysqlclient. Fairly sure depending on the client lib instead of a whole database server is the way to go here, though ;)

jtojnar commented on pkgs/servers/search/sphinxsearch/default.nix:
```diff
@@ -1,31 +1,44 @@
-    url = "http://sphinxsearch.com/files/sphinx-${version}-release.tar.gz";
-    sha256 = "1aa1mh32y019j8s3sjzn4vwi0xn83dwgl685jnbgh51k16gh6qk6";
-  }
+{ stdenv, fetchurl, pkgconfig, expat, libmysqlclient,
```
Should be used `pkg-config`

jtojnar commented on pkgs/servers/search/sphinxsearch/default.nix:
```diff
@@ -1,31 +1,44 @@
-  inherit version;
-  src = mainSrc;
+  version = "2.2.11";
+  src = fetchurl {
```
```suggestion

  src = fetchurl {
```

Valodim commented:
Thanks for the review @jtojnar! Fixed and tested compilation in all build configurations.

jtojnar commented:
@GrahamcOfBorg build sphinxsearch

vvs- reviewed (commented):
I guess there is no policy to prefer some particular database? This looks specific to NixOS as some other distributions I know always link to MariaDB for some reason.

Valodim committed:
sphinxsearch: add valodim as maintainer

Valodim committed:
sphinxsearch: add workaround for darwin build failures

upstream bug: see http://sphinxsearch.com/bugs/view.php?id=2578
fix ported from Homebrew: https://github.com/Homebrew/homebrew-core/pull/8619

Valodim commented:
Build failure on darwin :( seems like it's due to this upstream bug: http://sphinxsearch.com/bugs/view.php?id=2578

I imported a workaround for this particular issue from Homebrew:
https://github.com/Homebrew/homebrew-core/pull/8619/files#diff-df176d75fcc42a0ade34983e9ffc3193R47

I don't have a macOS to test on, unfortunately.

jtojnar commented on pkgs/servers/search/sphinxsearch/default.nix:
```diff
@@ -31,6 +31,14 @@ stdenv.mkDerivation rec {
+    "-Wno-reserved-user-defined-literal"
+    # workaround for "error: non-constant-expression cannot be narrowed from type 'long' to 'int'"
+    "-Wno-c++11-narrowing"
+  ];
```
Environment variables should be strings, see https://github.com/NixOS/nixpkgs/pull/72074

```suggestion
  CXXFLAGS = stdenv.lib.concatStringsSep " " (stdenv.lib.optionals stdenv.isDarwin [
    # see upstream bug: http://sphinxsearch.com/bugs/view.php?id=2578
    # workaround for "error: invalid suffix on literal
    "-Wno-reserved-user-defined-literal"
    # workaround for "error: non-constant-expression cannot be narrowed from type 'long' to 'int'"
    "-Wno-c++11-narrowing"
  ]);
```

Valodim replied:
whoops :)

jtojnar commented:
@GrahamcOfBorg build sphinxsearch

Valodim commented:
:tada:

jtojnar commented:
Thanks.
